### ISSUE TITLE
Service Entry is getting the incorrect workloads with selector

### ISSUE THREAD

**Issue by iandyh** [istio/istio#27935]
**Bug description**
A sample serviceentry

```
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: multi-cluster-demo
spec:
  hosts:
  - sample.global
  location: MESH_INTERNAL
  ports:
  - name: http
    number: 80
    protocol: http
  resolution: STATIC
  workloadSelector:
    labels:
      app: exporter-app
```
And have a deployment pointing to exporter-app. Restart the deployment couple of times you can see some stale endpoints lingering in the Envoy configurations. It does not happen after every restart but should be pretty easy to reproduce after couple of restarting. (The issue is gone after restarting the istiod)

Another observation: An unready pod could also be added to the outbound list, which should not happen. 

[ ] Docs
[ ] Installation
[X ] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Expected behavior**
ServiceEntry endpoint list should only contain serving pods that have passed the readiness check 

**Steps to reproduce the bug**
Described as above

**Version (include the output of `istioctl version --remote` and `kubectl version --short` and `helm version` if you used Helm)**
```
client version: 1.7.0
control plane version: 1.7.3
data plane version: 1.7.3 (10 proxies), 1.7.1 (18 proxies), 1.6.5 (23 proxies), 1.6.4 (57 proxies), 1.6.8 (28 proxies)
```

**How was Istio installed?**
Istio operator

**Environment where bug was observed (cloud vendor, OS, etc)**

**Comment by iandyh** [comment]
cc @howardjohn

**Comment by hzxuzhonghu** [comment]
I remember we have rounds of fixes for this

**Comment by iandyh** [comment]
@hzxuzhonghu yes, i was eager to test after our lab cluster upgraded to 1.7.3. However it was better compared to 1.6.x but the issues still exist.

**Comment by hzxuzhonghu** [comment]
OK, Thanks. I will investigate.

**Comment by hzxuzhonghu** [comment]
I can not reproduce, but find that pod in terminating status is still in endpoints. We should remove it.

**Comment by iandyh** [comment]
@hzxuzhonghu I can still reproduce it in 1.7.4. Have you tried to restart to selecting workloads multiple times? I can reproduce it after restart the deployment 3 times.

**Comment by hzxuzhonghu** [comment]
What i did is restarting the deployment.

**Comment by yingzhuivy** [comment]
we also encountered this on both istio 1.9 and 1.10. restarting the selecting workloads can sometimes result in stale endpoints, and restart istiod pod fix it. did you guys end up figuring out why? @iandyh @hzxuzhonghu

**Comment by yingzhuivy** [comment]
i can only reproduce this in prod environment, not test ones, maybe because in prod k8s, there are a lot of other services deploying that may take up the stale ip? maybe that's why you can not repro @hzxuzhonghu

**Comment by iandyh** [comment]
hmm, this should be fixed in 1.9 though. I actually provided a fix. Let me check it later.

**Comment by yingzhuivy** [comment]
thanks! could you link me to your fix? i was testing in 1.9.5 and 1.10.3 btw.

**Comment by yingzhuivy** [comment]
seems like it only happens when there is a k8s service with the same name and port with the serviceentry that's also selecting the pods
